src/bp/migrations/v12_11_0-1592325280-migrate_say_nodes.ts

```diff
@@ -11,79 +13,131 @@ const migration: Migration = {
-        }
-      })
-    }
+    const migrater = new Migrater(bp, inversify)
+    await migrater.initialize()
 
     if (metadata.botId) {
-      await updateFlowContent(metadata.botId)
+      await migrater.migrateBot(metadata.botId)
     } else {
       const bots = await bp.bots.getAllBots()
       for (const botId of Array.from(bots.keys())) {
-        await updateFlowContent(botId)
+        await migrater.migrateBot(botId)
       }
     }
 
     return {
       success: true,
-      message: hasChanged ? 'Content migrated successfully' : 'The content was already in the correct format'
+      message: migrater.hasChanged ? 'Content migrated successfully' : 'The content was already in the correct format'
     }
   }
 }
 
+class Migrater {
+  public hasChanged!: boolean
+  private flowService!: FlowService
+  private cmsService!: CMSService
+
+  constructor(private bp: typeof sdk, private inversify: Container) {}
+
+  async initialize() {
+    this.flowService = this.inversify.get<FlowService>(TYPES.FlowService)
+    this.cmsService = this.inversify.get<CMSService>(TYPES.CMSService)
+    await this.cmsService.initialize()
+  }
+
+  async migrateBot(botId: string) {
+    const ghost = this.bp.ghost.forBot(botId)
+    const flows = await this.flowService.loadAll(botId)
+
+    for (const flow of flows) {
+      try {
+        for (const node of flow.nodes) {
+          this.migrateNode(node)
+        }
+
+        const flowContent = {
+          ..._.pick(flow, ['version', 'catchAll', 'startNode', 'skillData', 'triggers', 'label', 'description']),
+          nodes: flow.nodes.map(node => _.omit(node, 'x', 'y', 'lastModified'))
+        }
+        await ghost.upsertFile('./flows', flow.location!, JSON.stringify(flowContent, undefined, 2))
+      } catch (err) {
+        this.bp.logger
+          .forBot(botId)
+          .attachError(err)
+          .error('Could not migrate say node data')
+      }
+    }
+  }
+
+  migrateNode(node: sdk.FlowNode) {
+    if (node.type !== 'say_something' || node.contents || !node['content']) {
+      return
+    }
+
+    const { contentType, formData } = (node as any).content
+    const { newSchema } = this.cmsService.getContentType(contentType)
+    const schema = <sdk.FormDefinition>newSchema
+
+    const fields = [...schema.advancedSettings, ...schema.fields]
+    const root: any = { formData: { contentType } }
+
+    for (const [key, value] of Object.entries(formData)) {
+      const [name, lang] = key.split('$')
+      this.attachTranslation(root, 'formData', { [name]: value }, fields, lang)
+    }
+
+    delete (node as any).content
+    node.contents = [root.formData]
+
+    this.hasChanged = true
+  }
+
+  attachTranslation(root: any, basePath: string, current: any, fields: sdk.FormField[], lang: string) {
+    for (const [key, value] of Object.entries<any>(current)) {
+      const field = fields.find(x => x.key === key)
+      if (!field) {
+        continue
```

Comment:
Just to be extra sure, if a field doesn't exist in the schema, leave the property in place instead of removing it